### PR TITLE
feat: issue #3589: support url signing for all valid elements in array

### PR DESCRIPTION
**Pull request by cameronbronstein**
I guess I'm not sure if the `feat` or `fix` flag is more appropriate here.

### PR fulfills these requirements
- [ ] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features) - _Please see below. I'm not sure where the best place is for a test._
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [x] Backend (Database)
- [ ] Backend (API)
- [ ] Frontend

### Describe the reason for change
https://github.com/heartexlabs/label-studio/issues/3589 - "Support URL signing when Image tag is an array of URIs"

#### What does this fix?
URL-signing is not supported for an array of URIs.

#### What is the new behavior?
If a value in task.data is an array and each element is a URI, signed URLs will be returned for each element.

#### What is the current behavior?
If a value in task.data is an array and each element in a URI, signed URLs cannot be produced, and the assets (e.g. images) are not viewable in the UI. 

#### What libraries were added/updated?
`io_storages.base_models`

#### Does this change affect performance?
It shouldn't have a noticeable affect on performance.

#### Does this change affect security?
No.

#### What alternative approaches were there?
I can't think of any. It's possible that the implementation should be moved to a different part of the code base. Pardon my lack of comprehensive knowledge here!

#### What feature flags were used to cover this change?
N/A

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit

I did a cursory glance but could not identify where a test should be implemented/modified. Please advise here.

### Which logical domain(s) does this change affect?
This effects cloud storage assets streaming/ import storage assets.

### CODE REVIEW

File: label_studio/io_storages/base_models.py [hunk 1]
```diff
@@ -76,12 +76,25 @@ def can_resolve_url(self, url):
...
+                    sub_uri, prefixes=(self.url_scheme,)
+                )
+                if not extracted_storage:
+                    logger.info(f'No storage info found for URI={sub_uri}')
+                    return
+
+                http_url = self.generate_http_url(extracted_uri)
+                resolved_uris.append(sub_uri.replace(extracted_uri, http_url))
+
+            if len(uri) == 1:
```
Comment (makseq) [inline label_studio/io_storages/base_models.py]: what if task.data = 
```
{ images: ["1.jpg"] }
```
this code will incorrectly return resolved_uris[0], however it must be resovled_uris. 

I would write something like this: 

```
def resolve_uri(self, uri):
  # process list of urls
  if isinstance(uri, list):
      resolved = []
      for sub in url:
          resolved.append(self.resolve_uri(sub))
      return resolved

  # process one url
  else:
        try:
            extracted_uri, extracted_storage = get_uri_via_regex(uri, prefixes=(self.url_scheme,))
            if not extracted_storage:
                logger.info(f'No storage info found for URI={uri}')
                return
            http_url = self.generate_http_url(extracted_uri)
            return uri.replace(extracted_uri, http_url)
        except Exception as exc:
            logger.info(f'Can\'t resolve URI={uri}', exc_info=True)

**Review by makseq — CHANGES_REQUESTED** [review]

File: label_studio/io_storages/base_models.py (hunk 1 above)
Comment (cameronbronstein) [reply label_studio/io_storages/base_models.py]: Thanks for the feedback. I agree with your suggestion.

Comment (makseq) [reply label_studio/io_storages/base_models.py]: Thank you for your changes. Now we should wait for tests from our QA team @borisheartex

**Comment by borisheartex** [comment]
/git merge develop
> Already up-to-date. Nothing to commit.
>
> [Workflow run](https://github.com/heartexlabs/label-studio/actions/runs/4041694406)

**Comment by cameronbronstein** [comment]
Hi @borisheartex @makseq . What's the current status here? Please let me know if there is anything I can do on my end to push this through. Thanks.

**Review by makseq — APPROVED** [review]

**Comment by cameronbronstein** [comment]
👍 🙇

**Comment by cameronbronstein** [comment]
@bmartel here's the issue I worked on that I mentioned in the meeting.

**Pull request merged**